### PR TITLE
feat(menu): enable having nested menus (sub-menus)

### PR DESCRIPTION
**Pull request by TommyLindh2**
This is one of three PRs that are connected and extracted from #2574 
They are chained in 3 branches and is best reviewed in this order:

1. https://github.com/Lundalogik/lime-elements/pull/2576 **This PR**
2. https://github.com/Lundalogik/lime-elements/pull/2577
3. https://github.com/Lundalogik/lime-elements/pull/2578

## Review:
- [x] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [x] Commits have the correct *type* for the changes made
- [x] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [x] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

File: src/components/menu-list/menu-list-renderer.tsx
Comment (Kiarokh) [inline src/components/menu-list/menu-list-renderer.tsx:255]: I'd change the commit title to `feat(menu): enable having nested menus (sub-menus)`

File: src/components/menu/menu.tsx
Comment (Kiarokh) [inline src/components/menu/menu.tsx:25]: https://github.com/Lundalogik/lime-elements/assets/35954987/5532be11-822b-492d-9b3d-20cb94e67fea

There is a bug in the keyboard navigation. When navigating the menu with keys, if I click outside to close the menu, then I can't open it again by clicking on it. (I think this happens if I have used the left and right arrow keys in my keyboard navigation)

Comment (adrianschmidt) [reply src/components/menu/menu.tsx:25]: Has this been fixed, or is this a call for help from the reviewer?

Comment (Kiarokh) [reply src/components/menu/menu.tsx:25]: No I just saw that this bug exists.

Comment (adrianschmidt) [reply src/components/menu/menu.tsx:25]: Ah, I understand. You've helped Tommy and Björn so much with this I stopped thinking of you as a reviewer and started thinking about you as one of the creators of the PR 🤦‍♂️ 😅

For me, it seems like it's pretty consistent that once I've opened the menu through keyboard navigation, regardless of how I closed it (clicking outside, clicking on a (leaf node) menu item, pressing escape, or pressing enter to select a (leaf node) menu item), I can't open the menu again by clicking the menu-button. If I tab to the menu button, I have to press enter twice to open the menu. This also seems consistent for me.

I've also tried opening the menu by clicking the button, and _then_ navigating with the arrow keys, and selecting a (leaf node) menu item, and then I can still open the menu by clicking it again. So there seems to be something that is mucked up by opening the menu using the keyboard. I would guess that there is some internal state that is set only when the menu is opened using the keyboard, and that is only reset by again trying to open the menu using the keyboard (which would explain why the first enter-press does nothing).

Comment (adrianschmidt) [reply src/components/menu/menu.tsx:25]: Further evidence: if I get into the bugged state, then tab to the button and press enter _once_, I can then open the menu by clicking on it, as usual.

File: src/components/menu/examples/menu-open-sub-menu-programatically.tsx
Comment (adrianschmidt) [inline src/components/menu/examples/menu-open-sub-menu-programatically.tsx:104]: I find it confusing that we have both `selectedItem` and `selectedItems` here.

The `selectedItems` array is, unless I'm mistaken, simply the items the menu should be showing at the moment. No one has selected them yet.

The name `selectedMenuItem` for the property that determines what sub-menu to show is also really confusing. Sure, I can see that you probably have selected that item to get to the state where the sub-menu is shown (unless the menu was opened programmatically), but I think a name like `submenuToDisplay` or `currentSubmenu` would be better. I know it's a little weird since that item doesn't really _contain_ the menu items shown, but that's what we convey to the user, so I think it's easier to grasp the code if the code is somewhat aligned with that concept.

File: src/components/menu/menu.tsx
Comment (TommyLindh2) [reply src/components/menu/menu.tsx:25]: After looking into it a bit it seems like what's causing this is the `makeEnterClickable` class that is supposed to prevent holding the Enter to produce a lot of click events, that is used for `limel-button`

Sometime the Up/Down events are triggered in the opposite order i.e. First "Up", then "Down", which will tell the click handler that the button is already "active" and should not let click events passed. So when you use Enter again, the up/down is triggered in the right order that will make it work again. Not sure how to fix it though.

Comment (TommyLindh2) [reply src/components/menu/menu.tsx:25]: I can re-create it in this docs as well: https://lundalogik.github.io/lime-elements/versions/NEXT-37.1.0-next.52/#/component/limel-menu/

File: src/components/menu/examples/menu-open-sub-menu-programatically.tsx
Comment (TommyLindh2) [reply src/components/menu/examples/menu-open-sub-menu-programatically.tsx:104]: Makes sense, will fix that

Comment (adrianschmidt) [reply src/components/menu/examples/menu-open-sub-menu-programatically.tsx]: Hmm… hang on… in the `CascadingMenuItems` imported from `item-constants.ts`, the sub-menu items actually do _contain_ the items included in that sub-menu. Why is it different here?

Another difference is that all the items in `selectedItems` have a reference to their `parentItem`, which apparently isn't necessary when not opening the menu programmatically?

The "vanilla" example for cascading menu is so clean, you just supply a nested structure of items, and everything is just taken care of, but here, there is so much extra stuff to do, you have to create new items yourself, and replace the normal structure of items with this specially crafted one, and then remember to restore the normal structure again at the right moments (plural!).

Have you explored the possibility of keeping the simple "rootItems" and just telling the menu where to navigate to, somehow? I imagine you could provide it with a "path". I guess since it's all arrays, and `value` is optional, rather than a nice `'format.lists'`, you'd give it something like `'0.2'` or `'[0][2]'`, but still, it would be much less work on the consumer.

By the way, I just noticed this little curiosity, which I think highlights why it would be nice to not have the consumer do all this "manual labour":

https://github.com/Lundalogik/lime-elements/assets/154725/e698622a-73bf-4d0e-abed-f3ca118aab6d

Since I feel this feedback might be a little hard to hear, considering it might require some rework, I wan't to express that the cascading menu feature in itself is absolutely fantastic, and you've really done some great work here!

Is the ability to programmatically open the menu to any sub-menu critical for your use-case? If not, I would suggest simply leaving that functionality out for now, and add the feature once it's ready and stable.

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [reply src/components/menu/menu.tsx:25]: I've created a new issue for this: https://github.com/Lundalogik/lime-elements/issues/2579

Resolving this thread.

File: src/components/menu/examples/menu-open-sub-menu-programatically.tsx
Comment (adrianschmidt) [inline src/components/menu/examples/menu-open-sub-menu-programatically.tsx:80]: I'm moving this comment since the thread it was in was resolved. Since the code has been updated, it's not describing the current state of the code 100 % correctly anymore, but the gist of it is still relevant:

---

Hmm… hang on… in the `CascadingMenuItems` imported from `item-constants.ts`, the sub-menu items actually do _contain_ the items included in that sub-menu. Why is it different here?

Another difference is that all the items in `selectedItems` have a reference to their `parentItem`, which apparently isn't necessary when not opening the menu programmatically?

The "vanilla" example for cascading menu is so clean, you just supply a nested structure of items, and everything is just taken care of, but here, there is so much extra stuff to do, you have to create new items yourself, and replace the normal structure of items with this specially crafted one, and then remember to restore the normal structure again at the right moments (plural!).

Have you explored the possibility of keeping the simple "rootItems" and just telling the menu where to navigate to, somehow? I imagine you could provide it with a "path". I guess since it's all arrays, and `value` is optional, rather than a nice `'format.lists'`, you'd give it something like `'0.2'` or `'[0][2]'`, but still, it would be much less work on the consumer.

By the way, I just noticed this little curiosity, which I think highlights why it would be nice to not have the consumer do all this "manual labour":

https://github.com/Lundalogik/lime-elements/assets/154725/e698622a-73bf-4d0e-abed-f3ca118aab6d

Since I feel this feedback might be a little hard to hear, considering it might require some rework, I wan't to express that the cascading menu feature in itself is absolutely fantastic, and you've really done some great work here!

Is the ability to programmatically open the menu to any sub-menu critical for your use-case? If not, I would suggest simply leaving that functionality out for now, and add the feature once it's ready and stable.

Comment (TommyLindh2) [reply src/components/menu/examples/menu-open-sub-menu-programatically.tsx:80]: Yeah, I know that it isn't a very good design on this specific feature, YES it is critical in our implementation of the hierarchy picker. But maybe we could solve it from the outside some how. The reason the "path thing" doesn't work is because when we later enable lazy-loading which we already have in the next PR, the items isn't arrays anymore but functions that load the data when clicked. Because the limel-menu doesn't have all the data of all hierarchies it will be hard to define. We've tried to simplify it but haven't find a good way.

I have updated the example so it fetches the items from the rootItems instead of defining them again, that might make the example a little bit easier to read. If you have any ideas on how to make it easier I'm all ears, or if you want to talk and discuss it.

Comment (adrianschmidt) [reply src/components/menu/examples/menu-open-sub-menu-programatically.tsx:80]: Ah, I understand. Yeah that seems like it would make the path approach more difficult.

Then I would suggest that we leave the functionality in, but actually remove the example, and add something to the docs about `selectedMenuItem` being an experimental property which shouldn't be used, and which we may remove without warning in the future. Then _you_ can use it, but we don't promise anyone else we'll keep it, if we find a better solution. And if we don't, we can make it permanent and add the example back at a later time.

Sound ok?

Comment (TommyLindh2) [reply src/components/menu/examples/menu-open-sub-menu-programatically.tsx:80]: Yeah, sounds like a plan.
Then we can make the `parentItem` property on `MenuItem` private as well.

**Comment by TommyLindh2** [comment]
@adrianschmidt why did you change the base to `next`? I created a separate branch to merge all these three so that we can test it all together in the end and not release parts of the changes in the menu but everything all at once.

File: src/components/menu/examples/menu-open-sub-menu-programatically.tsx
Comment (TommyLindh2) [reply src/components/menu/examples/menu-open-sub-menu-programatically.tsx:82]: Done, make sure to verify that you agree with the changes, and if so, Resolve the conversation :)

**Comment by adrianschmidt** [comment]
> why did you change the base to `next`? I created a separate branch to merge all these three so that we can test it all together in the end and not release parts of the changes in the menu but everything all at once.

Why does it need to be released all at once? We generally want to get each mergeable part merged as soon as possible, to avoid having giant PRs and risk massive merge conflicts that can be very difficult to resolve. The longer a branch stays unmerged, the greater the risk of conflicting changes being developed and merged, causing even further delays, increasing the risk of even more conflicting changes. We want all development to happen on branches that remain as small as possible and as short-lived as possible.

Are there changes in the upcoming PRs that will be breaking if this is merged first?

**Comment by TommyLindh2** [comment]
> > why did you change the base to `next`? I created a separate branch to merge all these three so that we can test it all together in the end and not release parts of the changes in the menu but everything all at once.
> 
> Why does it need to be released all at once? We generally want to get each mergeable part merged as soon as possible, to avoid having giant PRs and risk massive merge conflicts that can be very difficult to resolve. The longer a branch stays unmerged, the greater the risk of conflicting changes being developed and merged, causing even further delays, increasing the risk of even more conflicting changes. We want all development to happen on branches that remain as small as possible and as short-lived as possible.
> 
> Are there changes in the upcoming PRs that will be breaking if this is merged first?

No I guess not, maybe it's just because they all belong together in my head. But there shouldn't be any issue releasing them one-by-one

File: src/components/menu/menu.types.ts
Comment (TommyLindh2) [inline src/components/menu/menu.types.ts]: It would be weird if the didn't but yes, there's nothing saying that they must.

File: src/components/menu/examples/menu-open-sub-menu-programmatically.tsx
Comment (TommyLindh2) [inline src/components/menu/examples/menu-open-sub-menu-programmatically.tsx:34]: I really like this rename. We didn't come up with a good name.
But you did ❤️

File: src/components/menu/examples/menu-open-sub-menu-programmatically.tsx
Comment (adrianschmidt) [inline src/components/menu/examples/menu-open-sub-menu-programmatically.tsx:69]: We don't need to change anything here, since we're currently not using this example anyway, but we don't want people to use arrow-functions or `.bind()` in templates, because both of them cause a new function to be created every time the template is re-rendered, which eats up resources and performance. And since we don't want people to do that, we also really don't want to teach them to do it by doing it in the code examples 😊

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [inline src/components/menu/menu.tsx:389]: This doesn't need to be async yet. I'll fix 👍

File: src/components/menu/menu.types.ts
Comment (adrianschmidt) [inline src/components/menu/menu.types.ts:85]: It seems like the change in type of this property from `Array<MenuItem<T> | ListSeparator>` to `Array<MenuItem<T> | ListSeparator> | MenuLoader` (in the next PR) might actually be a breaking change* if we don't release those PRs together, so I'm holding off on merging this for a little bit anyway.

*) Evidence: see the change I had to make to the "open sub-menu programmatically"-example in the next PR: https://github.com/Lundalogik/lime-elements/pull/2577/files#diff-2add54a9e21a898b280c548832b0213ea57812aa63e2a29aecabbc0afd6389e8R87

image

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [inline src/components/menu/menu.tsx:355]: I don't understand why we're going to "grandParent" here? If we want to navigate one step up from "current", surely that must be "parent" then? 🤔

**Review by adrianschmidt — COMMENTED** [review]
Oh, apparently I "started a review". It's so annoying that that's the default option when commenting 😣

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [inline src/components/menu/menu.tsx:230]: Having to redo this loop on every render feels a little painful. I'm not saying we necessarily need to optimise it _now_, but its good to be aware of whenever you are in any piece of code that is part of the render-loop, as that means the code will be run pretty incessantly, and will have a disproportionate performance impact.

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [inline src/components/menu/menu.tsx:241]: Styles belong in the stylesheet 🙂

Comment (adrianschmidt) [reply src/components/menu/menu.tsx:241]: Oh, or… these can't be in the stylesheet, because this stuff is lifted out to the portal, right?

That brings me to the point that this seems complex enough that it might warrant breaking it out into its own (internal) component. And it just so happens that doing so would let us avoid the inline styles as well!

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [inline src/components/menu/menu.tsx:264]: If we optimise the breadcrumbs-creating loop so it doesn't have to be run on every single render, we should move this out too, for the same reason.

We should be able to update the breadcrumbs when navigation happens, and just keep that state between renders.

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [reply src/components/menu/menu.tsx:355]: Ah, is it because we we're not going to the parent, we're selecting the grandparent item, which den loads the parent? Is that also why there is loading of items when moving outwards from a sub-menu to its parent menu (in the lazy-loading example in the next PR)?

**Comment by adrianschmidt** [comment]
I'm going to log off for today after writing this comment, and perhaps it will feel different after resting my brain until tomorrow, but I'm worried about how complex this whole thing is. Like, what's in the first PR is fairly manageable, but adding in the lazy-loading it's starting to feel like it's melting my brain, and I haven't even gotten to the search-functionality yet.

Don't get me wrong, the functionality provided to the user is _excellent_! The problem I see is how are we ever going to maintain this code? Who's going to dare to make any changes to it?

I have a bunch of more or less vague ideas for how things might potentially be possible to simplify, but I'm not sure if they'll work or not, and I feel awful about once again being such an absolute pita before something you built can be merged 😭

File: src/components/menu/menu.tsx
Comment (TommyLindh2) [reply src/components/menu/menu.tsx:355]: Yes, exactly. We "simulate" a click on the grandParent to go to the parent so to say.

File: src/components/menu/menu.tsx
Comment (TommyLindh2) [reply src/components/menu/menu.tsx:264]: Yeah, before the logic for this was very different, but now when we have the `currentSubMenu` in a prop it's easier to do that. I can try it out.

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [reply src/components/menu/menu.tsx:271]: > The failing test seems to be failing because of this tabindex. (Although, I tried running the tests locally, both with this left in, and with it removed, and all tests passed regardless 🤷‍♂️)

Ah, no, I tried re-running the tests now, and they passed, so it was just a timing-thing or something I guess. The tests can be a little bit flaky sometimes… 😒

I'd still like to know what this does, or remove it 🙂

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [reply src/components/menu/menu.tsx]: ⚡

Comment (adrianschmidt) [reply src/components/menu/menu.tsx]: Or, well, no, not quite zapped. I've just improved it a little bit. It still runs on every render.

Comment (adrianschmidt) [reply src/components/menu/menu.tsx]: Sorry, I'm too tired and don't read properly. This is indeed zapped, because, while the loop is still run on every render, this specific thing has been moved, so it's done next to the loop. So I'll resolve this after all. Sorry for spamming… 🙇‍♂️

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [reply src/components/menu/menu.tsx]: ⚡

File: src/components/menu/examples/menu-open-sub-menu-programmatically.tsx
Comment (TommyLindh2) [reply src/components/menu/examples/menu-open-sub-menu-programmatically.tsx:69]: We add the file, so I think it's a good idea to fix it anyways. It's quick.

File: src/components/menu/menu.tsx
Comment (TommyLindh2) [reply src/components/menu/menu.tsx:248]: Yeah, me and Kia discussed that, but it would only be a very small wrapper of the limel-breadcrumbs which seems kind of risky for only 2 style properties.

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [inline src/components/menu/menu.tsx]: This change is to avoid any potential risk of getting out of sync.

File: src/components/menu/menu.tsx
Comment (adrianschmidt) [reply src/components/menu/menu.tsx]: We can remove it for now. It might be needed once search is added in a later PR.

**Review by adrianschmidt — APPROVED** [review]

**Comment by Kiarokh** [comment]
😮 Merged! Wow... yeay....! 🥳 

Great job guys! ❤️

**Comment by TommyLindh2** [comment]
It's only nested menu, without Search and lazy loading. So a little more work needs to be done :)

**Pull request merged**
